src/libs/kata-types/src/config/hypervisor/qemu.rs

```diff
@@ -0,0 +1,146 @@
+// Copyright (c) 2019-2021 Alibaba Cloud
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use std::io::Result;
+use std::path::Path;
+use std::sync::Arc;
+
+use super::{default, register_hypervisor_plugin};
+
+use crate::config::default::MAX_QEMU_VCPUS;
+use crate::config::default::MIN_QEMU_MEMORY_SIZE;
+
+use crate::config::hypervisor::VIRTIO_BLK_MMIO;
+use crate::config::{ConfigPlugin, TomlConfig};
+use crate::{eother, resolve_path, validate_path};
+
+/// Hypervisor name for qemu, used to index `TomlConfig::hypervisor`.
+pub const HYPERVISOR_NAME_QEMU: &str = "qemu";
+
+/// Configuration information for qemu.
+#[derive(Default, Debug)]
+pub struct QemuConfig {}
+
+impl QemuConfig {
+    /// Create a new instance of `QemuConfig`.
+    pub fn new() -> Self {
+        QemuConfig {}
+    }
+
+    /// Register the qemu plugin.
+    pub fn register(self) {
+        let plugin = Arc::new(self);
+        register_hypervisor_plugin(HYPERVISOR_NAME_QEMU, plugin);
+    }
+}
+
+impl ConfigPlugin for QemuConfig {
+    fn get_max_cpus(&self) -> u32 {
+        MAX_QEMU_VCPUS
+    }
+
+    fn get_min_memory(&self) -> u32 {
+        MIN_QEMU_MEMORY_SIZE
+    }
+    fn name(&self) -> &str {
+        HYPERVISOR_NAME_QEMU
+    }
+
+    /// Adjust the configuration information after loading from configuration file.
+    fn adjust_configuration(&self, conf: &mut TomlConfig) -> Result<()> {
+        if let Some(qemu) = conf.hypervisor.get_mut(HYPERVISOR_NAME_QEMU) {
+            if qemu.path.is_empty() {
+                qemu.path = default::DEFAULT_QEMU_BINARY_PATH.to_string();
+            }
+            resolve_path!(qemu.path, "Qemu binary path `{}` is invalid: {}")?;
+            if qemu.ctlpath.is_empty() {
+                qemu.ctlpath = default::DEFAULT_QEMU_CONTROL_PATH.to_string();
+            }
+            resolve_path!(qemu.ctlpath, "Qemu ctlpath `{}` is invalid: {}")?;
+
+            if qemu.boot_info.kernel.is_empty() {
+                qemu.boot_info.kernel = default::DEFAULT_QEMU_GUEST_KENREL_IMAGE.to_string();
+            }
+            if qemu.boot_info.kernel_params.is_empty() {
+                qemu.boot_info.kernel_params =
+                    default::DEFAULT_QEMU_GUEST_KENREL_PARAMS.to_string();
+            }
+            if qemu.boot_info.firmware.is_empty() {
+                qemu.boot_info.firmware = default::DEFAULT_QEMU_FIRMWARE_PATH.to_string();
+            }
+
+            if qemu.device_info.default_bridges == 0 {
+                qemu.device_info.default_bridges = default::DEFAULT_QEMU_PCI_BRIDGES;
+                if qemu.device_info.default_bridges > default::MAX_QEMU_PCI_BRIDGES {
+                    qemu.device_info.default_bridges = default::MAX_QEMU_PCI_BRIDGES;
+                }
```

Comment:
This isn't really necessary fter setting the default IMO.

Reply:
fixed